### PR TITLE
V3 html

### PR DESCRIPTION
**Pull request by jonludlam**
This PR adds a few options to the HTML renderer - to omit breadcrumbs, to omit the table of contents, to output only the content rather than within a complete html page, and to allow the table of contents to be output to a json file. This is what's required to allow the output to be used directly by v3.ocaml.org rather than having to use the forked renderer in voodoo.git.

As well as being useful for v3.ocaml.org, this should be useful to anyone who wants to embed the output of odoc as a component within a larger page.

This is just a draft right now, it'd be useful to get some feedback on whether these are the right set of options, whether bringing in yojson as a normal rather than test dependency is a good idea and so on.

### CODE REVIEW

**Comment by dbuenzli** [comment]
How does that fare to what was discussed in https://github.com/ocaml/odoc/issues/374 ? 

> whether bringing in yojson as a normal rather than test dependency is a good idea and so on.

Please don't. The only good idea as far as `odoc` is concerned is to [remove](https://github.com/ocaml/odoc/pull/305) [dependencies](https://github.com/ocaml/odoc/pull/306). You are simply generating json, that's trivial, you just have to escape strings properly. Here's a function that will add a valid json string to a buffer:

```ocaml
let buffer_add_json_string b s =
  let is_control = function '\x00' .. '\x1F' | '\x7F' -> true | _ -> false in
  let len = String.length s in
  let max_idx = len - 1 in
  let flush b start i =
    if start < len then Buffer.add_substring b s start (i - start);
  in
  let rec loop start i = match i > max_idx with
  | true -> flush b start i
  | false ->
      let next = i + 1 in
      match String.get s i with
      | '"' -> flush b start i; Buffer.add_string b "\\\""; loop next next
      | '\\' -> flush b start i; Buffer.add_string b "\\\\"; loop next next
      | c when is_control c ->
          flush b start i;
          Buffer.add_string b (Printf.sprintf "\\u%04X" (Char.code c));
          loop next next
      | c -> loop start next
  in
  Buffer.add_char b '"'; loop 0 0; Buffer.add_char b '"';
```

**Comment by Drup** [comment]
> This PR adds a few options to the HTML renderer - to omit breadcrumbs, to omit the table of contents, to output only the content rather than within a complete html page

This seems useful, but since all of these are sort of related, I would suggest using a constructed type indicating what should be in the output.

As for the json output ... It was my understanding that the general opinion was against "outputing things on the side", so shouldn't we have a different mode that does only the JSON thingy ?

**Comment by jonludlam** [comment]
@Drup :
> It was my understanding that the general opinion was against "outputing things on the side", so shouldn't we have a different mode that does only the JSON thingy ?

I think the important thing here is that we're outputting metadata that's specific to the HTML backend, so I'm quite keen to keep it as part of the backend. If we need some more generic metadata we might end up sticking that in its own backend.

> constructed type indicating what should be in the output

I'm not quite sure what you're suggesting. I did add a type gathering all the options together [here](https://github.com/ocaml/odoc/pull/842/files#diff-74c948bfdafcf27037ab5e118cb70bcdc73992cd5e239f7a2cd198b25130456aR3-R14) - did you mean more than that?

@dbuenzli :
> Here's a function that will add a valid json string to a buffer:

Hrm - it's not hard, but it's also nothing to do with odoc, and on that basis I'm not particularly keen to maintain that sort of function. Having said that, I do take your point about adding dependencies (as I said in my first comment), so I'm not ruling it out. One further suggestion then: how about changing the output to a sexp and using `sexplib0`? It will add no extra dependencies beyond that small library?

> How does that fare to what was discussed in https://github.com/ocaml/odoc/issues/374 ?

It goes a bit beyond what was suggested there (by making the ToC and breadcrumbs optional too). We found it relatively useful to _not_ change the filename, as the output will still render in the browsers I tried and also the links actually still work. I can be persuaded otherwise though. I also steered clear of the `--fragment` thing as it doesn't do quite the same thing as `html-fragment`, but again I'm open to persuasion if anyone has strong views on that.

**Comment by Drup** [comment]
> I'm not quite sure what you're suggesting. I did add a type gathering all the options together [here](https://github.com/ocaml/odoc/pull/842/files#diff-74c948bfdafcf27037ab5e118cb70bcdc73992cd5e239f7a2cd198b25130456aR3-R14) - did you mean more than that?

```
type output = Page { toc : bool ; breadcrump : bool } | Fragment
```

Something like that ? I guess there are others that can be added. In any case, I'm not fond of having collection of booleans which are not orthogonal.

**Comment by trefis** [comment]
Heya,

If you'll allow me, I'm going to make the obvious suggestion of splitting that PR in 2 (at least).
I understand why it was born that way, but given that parts of it seem more easy to accept than others, keeping it whole is likely to delay integration for a while.

A natural way to split it is to move all the json bits to their own self contained PR, where we can discuss json vs sexp vs xml, yojson vs sexplib vs internal stuff, etc.
And have everything else in another PR.

That hopefully shouldn't be too much work to do.

---

Then we can spend an unreasonable amount of time discussing the way options should be represented, and I'll be happy to disagree with @Drup. Because I obviously don't see why you might want to let users not have the breadcrumbs when they generate a whole page, but force them to have the breadcrumbs when they only generate a fragment.

**Comment by trefis** [comment]
> I'm not fond of having collection of booleans which are not orthogonals.

I think Jon sees them as orthogonal, and I share his view.

File: src/html/tree.ml [hunk 1]
```diff
@@ -16,23 +16,49 @@
...
+  and sections the_sections =
+    the_sections
+    |> List.map (fun the_section -> Html.li (section the_section))
+    |> Html.ul
+  in
+  match toc with
+  | [] -> []
+  | _ -> [ Html.nav ~a:[ Html.a_class [ "odoc-toc" ] ] [ sections toc ] ]
+
+let _json_of_toc toc =
```
Comment (trefis) [inline src/html/tree.ml:33]: Same here, let's remove that from this PR.

**Review by trefis — COMMENTED** [review]
I gave a first quick look at this.
There's still some json left here and there, but apart from that things seemed OK.

I'll try and do a more serious review shortly.

File: src/html/dune [hunk 2]
```diff
@@ -3,4 +3,4 @@
  (public_name odoc.html)
  (instrumentation
   (backend bisect_ppx))
- (libraries odoc_model odoc_document tyxml))
```
[the remark this answers is not in the document]
Comment (trefis) [reply src/html/dune:6]: (ah, of course the comment is poorly placed: the next line of the diff adds `yojson` to the libraries field.)

**Pull request merged**